File: p01ab.hs
-- Advent of Code 2024 - Day 1 part One and Two
-- Solutions in Haskell
-- (Ter leering ende vermaeck...)
--
-- The total distance between the lists:     2031679
-- Their similarity score:                  19678534
--
-- (cl) by Arno Jacobs, 2024-12-01

module AoC2024d01ab where

import Data.Char (isDigit)
import Data.List (sort)

-- Some initials
filename :: String
filename = "data/inputDay01_2024.txt"


parseNumbers :: [String] -> ([Int],[Int])
parseNumbers = twoSortedLists . map (\xs -> (numberLeft xs, numberRight xs)) 
    where 
        numberLeft  = read . takeWhile isDigit
        numberRight = read . dropWhile isDigit
    --  twoSortedLists :: [(Int,Int)] -> ([Int],[Int])
        twoSortedLists ps = (sortedLeft ps,sortedRight ps)
            where
                sortedLeft  = sort . map fst
                sortedRight = sort . map snd

partOne :: [String] -> Int
partOne = sum . deltaLists . parseNumbers

deltaLists :: ([Int],[Int]) -> [Int]
deltaLists (xs,ys) = map (\(x,y) -> abs(x-y)) $ zip xs ys

partTwo :: [String] -> Int
partTwo = sum . tallyLists . parseNumbers

tallyLists :: ([Int],[Int]) -> [Int]
tallyLists (xs,ys) = map tally xs      
    where
        tally x = x * length (filter (==x) ys)


main :: IO ()
main = do   putStrLn "Advent of Code 2024 - day 1  (Haskell)"
            day1 <- lines <$> readFile filename
            putStr "Part one: The total distance between the lists:  "
            print $ partOne day1
            putStr "Part two: Their similarity score:               "
            print $ partTwo day1            
            putStrLn "0K.\n"


-- End of code

File: p01ab.clj
;;;;    Advent of Code 2024 - Day 1 part One and Two
;;;;    https://adventofcode.com/2024/day/1
;;;;
;;;;    Solutions in Clojure
;;;;    (Ter leering ende vermaeck...)
;;;;
;;;;    The total distance between the lists:     2031679
;;;;    Their similarity score:                  19678534
;;;;
;;;;    (cl) by Arno Jacobs, 2024-12-04
;;;;    

(ns p01ab
  (:require [clojure.string :as str]))

;;; The function "merge-sort" is in this file
(load-file "caoc/caoc.clj")

;;;; Read the data-set
(defn get-lines 
  [file]
  (->> file
       slurp
       (str/split-lines)))

(defn parse-line 
  [line]
  (mapv parse-long (str/split (str/trim line) #" ")))

(defn first-number
  [line]
  (->> line
       parse-line
       first))

(defn last-number 
  [line]
  (->> line
       parse-line
       last))

(def data-set (get-lines "./data/inputDay01_2024.txt"))
(def list1 (caoc/merge-sort (mapv first-number data-set)))
(def list2 (caoc/merge-sort (mapv last-number data-set)))


;;;  Part one

(defn total-distances
  [xs ys]
  (cond (empty? xs) 0
        (empty? ys) 0
        :else (let [d (abs (- (first xs) (first ys)))]
    (+ d (total-distances (rest xs) (rest ys))))))


;;; Part two

(defn tallyX
  [x ys]
  (* x (count (filter #(= x %) ys))))

(defn similarity-score
  [xs ys]
  (cond (empty? xs) 0
        :else (+ (tallyX (first xs) ys) (similarity-score (rest xs) ys))))


;;; The 'main' program - - -
(defn program []
  (println "Advent of Code 2024 - day 1  (Clojure)")
  (print   "Part one: The total distance between the lists:  ")
  (println (total-distances list1 list2))
  (print   "Part two: Their similarity score:               ")
  (println (similarity-score list1 list2))
  (println "0K.\n"))

;; Run in terminal via: clojure -M p01ab.clj
(program)

